compaction/compaction_manager.cc

```diff
@@ -254,121 +254,123 @@ compaction_manager::compaction_state& compaction_manager::get_compaction_state(r
         });
-    }).then_wrapped([this, task] (future<> f) {
-        _stats.active_tasks--;
-        _tasks.remove(task);
-        cmlog.debug("Major compaction {}: done", *task);
-        try {
-            f.get();
-            _stats.completed_tasks++;
-        } catch (sstables::compaction_stopped_exception& e) {
-            cmlog.info("major compaction stopped, reason: {}", e.what());
-        } catch (...) {
-            cmlog.error("major compaction failed, reason: {}", std::current_exception());
-            _stats.errors++;
-        }
-    });
-    return task->compaction_done().then([task] {});
+        finish_compaction();
+    }
+};
+
+future<> compaction_manager::perform_major_compaction(replica::table* t) {
+    if (_state != state::enabled) {
+        return make_ready_future<>();
+    }
+    return perform_task(make_shared<major_compaction_task>(*this, t));
 }
 
 class compaction_manager::custom_compaction_task : public compaction_manager::task {
+    noncopyable_function<future<>(compaction_manager::task&, sstables::compaction_data&)> _job;
+
 public:
-    custom_compaction_task(compaction_manager& mgr, replica::table* t, sstables::compaction_type type, sstring desc)
+    custom_compaction_task(compaction_manager& mgr, replica::table* t, sstables::compaction_type type, sstring desc, noncopyable_function<future<>(compaction_manager::task&, sstables::compaction_data&)> job)
         : task(mgr, t, type, std::move(desc))
+        , _job(std::move(job))
     {}
+
+protected:
+    virtual future<> do_run() override {
+        if (!can_proceed(throw_if_stopping::yes)) {
+            co_return;
+        }
+        switch_state(state::pending);
+        auto units = co_await get_units(_cm._maintenance_ops_sem, 1);
+        // take read lock for table, so major compaction and resharding can't proceed in parallel.
+        auto lock_holder = co_await _compaction_state.lock.hold_read_lock();
+        if (!can_proceed(throw_if_stopping::yes)) {
+            co_return;
+        }
+        setup_new_compaction();
+
+        // NOTE:
+        // no need to register shared sstables because they're excluded from non-resharding
+        // compaction and some of them may not even belong to current shard.
+        co_await _job(*this, _compaction_data);
+        finish_compaction();
+    }
 };
 
-future<> compaction_manager::run_custom_job(replica::table* t, sstables::compaction_type type, const char* desc, noncopyable_function<future<>(sstables::compaction_data&)> job) {
+future<> compaction_manager::run_custom_task_job(replica::table* t, sstables::compaction_type type, const char* desc, noncopyable_function<future<>(compaction_manager::task& task, sstables::compaction_data&)> job) {
     if (_state != state::enabled) {
         return make_ready_future<>();
     }
+    return perform_task(make_shared<custom_compaction_task>(*this, t, type, desc, std::move(job)));
+}
```

Comment:
@bhalevy I think we should merge run_custom_task_job() into run_custom_job(). It's better not to expose compaction_manager::task in any public interface, and it's also not needed anywhere. Let's only leave run_custom_job().

Reply:
fixed